days/src/day9/Day9.kt

```diff
@@ -0,0 +1,55 @@
+package day9
+
+import Day
+import java.io.File
+
+class Day9(private val xmasCode: List<Long>) : Day {
+    private val preambleLength = 25
+
+    private fun MutableList<Long>.isValidXMAS(num: Long)= this.asSequence().associateBy { num - it }
+            .let { adjusted -> this.any { adjusted.contains(it) } }
+
+    private fun List<Long>.findNonXmas(preambleLength: Int) =
+        this.take(preambleLength).toMutableList().let { window ->
+            this.drop(preambleLength).filterNot { current ->
+                window.isValidXMAS(current).also {
+                    window.removeAt(0)
+                    window.add(current)
+                }
+            }.first()
+        }
+    override fun puzzle1(): Int? = xmasCode.findNonXmas(preambleLength).toInt()
+
+    override fun puzzle2(): Int? {
+        return xmasCode.findNonXmas(preambleLength).let { target ->
+            var window = 0
+            val sum = xmasCode.foldIndexed(0L) { i, sum, current ->
+                val curr = if (current > target) {
+                    window = i + 1
+                    0L
+                } else if (sum + current > target) {
+                    var currentSum = sum + current
+                    while (currentSum > target) {
+                        currentSum -= xmasCode[window]
+                        window++
+                    }
+                    currentSum
+                } else {
+                    sum + current
+                }
```

Comment:
Yeah, my solution is similar to this but a bit simpler to read because I move both the start and end indexes (so I don't have the inner while loop. I do a computationaly expensive step of re-summing the elements at the end of each step (using a reduce) but that could easily be swapped to manual book keeping the way you do it here.